File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        def dfs(root, maxVal):
            if not root:
                return 0

            count = 0
            
            if root.val >= maxVal:
                count += 1

            count += dfs(root.left, max(maxVal, root.val)) + dfs(root.right, max(maxVal, root.val))

            return count
            
        return dfs(root, root.val)

File: Data Structures & Algorithms/same-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // if (p == null && q == null) return true;

        // if (p != null && q != null && p.val == q.val) {
        //     return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
        // }
        
        // return false;
        Queue<TreeNode> q1 = new LinkedList<>();
        Queue<TreeNode> q2 = new LinkedList<>();

        q1.add(p);
        q2.add(q);

        while (!q1.isEmpty() && !q2.isEmpty()) {
            for (int i = 0; i < q1.size(); i++) {
                TreeNode cur1 = q1.poll();
                TreeNode cur2 = q2.poll();

                if (cur1 == null && cur2 == null) continue;

                if (cur1 == null || cur2 == null || cur1.val != cur2.val) {
                    return false;
                }

                q1.add(cur1.left);
                q1.add(cur1.right);
                q2.add(cur2.left);
                q2.add(cur2.right);
            }
        }

        return true;
    }
}
